### PR TITLE
Apply class weights in OBB loss

### LINKED ISSUE #24841: cls_pw (class-imbalance weighting) has no effect on OBB training — v8OBBLoss omits the class_weights multiply
**Version:** ultralytics 8.4.37 · torch 2.11.0+cu130

### Summary

Setting `cls_pw` for an OBB task has no effect on training. The class weights are computed and logged at train start, the detection loss applies them, but `v8OBBLoss` reimplements the classification-loss term and never multiplies by `class_weights` — so the parameter is silently inert for `task=obb`.

### Evidence — the divergence is one line

In `ultralytics/utils/loss.py`:

`v8DetectionLoss` applies the weights:
```python
self.class_weights = getattr(model, "class_weights", None)
...
bce_loss = self.bce(pred_scores, target_scores.to(dtype))
if self.class_weights is not None:
    bce_loss *= self.class_weights          # ← applied
loss[1] = bce_loss.sum() / target_scores_sum
```

`v8OBBLoss` computes BCE without it:
```python
loss[1] = self.bce(pred_scores, target_scores.to(dtype)).sum() / target_scores_sum
#         ← no class_weights multiply
```

`DetectionTrainer.set_class_weights()` still runs for OBB (it logs `Class weights: [...]`), so the user sees weighting "active" while it has zero effect on gradients.

### Reproduction (deterministic)

Train the same OBB dataset 3× with `cls_pw=0.0`, `0.5`, `1.0` and a fixed `seed`. Expected: 3 different models. Actual: **bit-identical** weights and metrics (we observed identical holdout mAP50/recall/precision to 4 decimals across all three).

### Impact

OBB datasets with class imbalance can't use the documented imbalance lever; rare classes stay under-learned while the knob misleads. (Our case: a ~14× instance spread across look-alike sibling classes.)

### Suggested fix

Mirror the detect path — apply `class_weights` to the OBB BCE term, guarded by the same `if self.class_weights is not None`. Happy to open a PR with that one-line change plus a regression test if that's welcome.

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #24841

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
✨ This PR adds optional class-weight support to the classification loss, allowing training to better handle imbalanced datasets.

### 📊 Key Changes
- Added an intermediate `bce_loss` variable for classification loss calculation in `ultralytics/utils/loss.py`.
- Introduced a check for `self.class_weights` before finalizing the BCE classification loss.
- When class weights are provided, the per-class BCE loss is multiplied by those weights before summing and normalizing.
- Left the rest of the loss pipeline unchanged, including box loss behavior.

### 🎯 Purpose & Impact
- 🎯 Helps models learn more fairly from datasets where some classes appear much more often than others.
- ⚖️ Gives less frequent classes a stronger training signal, which can improve performance on underrepresented categories.
- 🧩 Makes the loss function more flexible without changing behavior for users who do not set class weights.
- 🚀 Useful for custom training workflows where class imbalance is a common challenge, with minimal impact on existing setups.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review, and an engineer will assist you shortly. Please review the checklist below for smooth integration 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, low-risk change. The OBB classification path now mirrors the existing class-weight handling used elsewhere in the loss code, and I don't see any correctness or maintainability issues in the lines changed.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Thanks, @Y-T-G, for this thoughtful improvement to classification loss handling.

As Marie Curie said, *“Nothing in life is to be feared, it is only to be understood.”* This PR reflects that spirit by giving users a clearer, more flexible way to handle class imbalance and help models learn more fairly from real-world datasets.

Really appreciate the clean, targeted implementation and the care taken to preserve existing behavior while expanding training flexibility.
